sql/src/main/java/io/crate/operation/merge/RamAccountingPageIterator.java

```diff
@@ -0,0 +1,116 @@
+        PositionalOrderBy orderBy = mergePhase.orderByPositions();
+
+        PagingIterator<Integer, Row> delegatePagingIterator =
+            getDelegatePagingIterator(mergePhase, requiresRepeat, orderBy);
+        if (requiresRepeat) {
+            return new RamAccountingPageIterator(
+                delegatePagingIterator,
+                new RowAccounting(
+                    mergePhase.inputTypes(),
+                    RamAccountingContext.forExecutionPhase(circuitBreaker, mergePhase)));
+        }
+        return delegatePagingIterator;
+    }
+
+    @VisibleForTesting
+    public static PagingIterator<Integer, Row> create(MergePhase mergePhase, RowAccounting rowAccounting) {
+        PositionalOrderBy orderBy = mergePhase.orderByPositions();
+        PagingIterator<Integer, Row> delegatePagingIterator =
+            getDelegatePagingIterator(mergePhase, true, orderBy);
+
+        return new RamAccountingPageIterator(delegatePagingIterator, rowAccounting);
+    }
+
+    private static PagingIterator<Integer, Row> getDelegatePagingIterator(MergePhase mergePhase,
+                                                                          boolean requiresRepeat,
+                                                                          PositionalOrderBy orderBy) {
+        PagingIterator<Integer, Row> delegatePagingIterator;
+        if (mergePhase.numUpstreams() == 1 || orderBy == null) {
+            if (requiresRepeat) {
+                delegatePagingIterator = PassThroughPagingIterator.repeatable();
+            } else {
+                delegatePagingIterator = PassThroughPagingIterator.oneShot();
+            }
+        } else {
+            delegatePagingIterator = new SortedPagingIterator<>(OrderingByPosition.rowOrdering(orderBy), requiresRepeat);
+        }
+        return delegatePagingIterator;
+    }
+
+    @Override
+    public void merge(Iterable<? extends KeyIterable<Integer, Row>> keyIterables) {
+        for (KeyIterable<Integer, Row> iterable : keyIterables) {
+            for (Row row : iterable) {
+                rowAccounting.accountForAndMaybeBreak(row);
+            }
+        }
+        delegatePagingIterator.merge(keyIterables);
+    }
+
+    @Override
+    public void finish() {
+        rowAccounting.close();
+        delegatePagingIterator.finish();
+    }
+
+    @Override
+    public Integer exhaustedIterable() {
+        return delegatePagingIterator.exhaustedIterable();
+    }
+
+    @Override
+    public Iterable<Row> repeat() {
+        return delegatePagingIterator.repeat();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return delegatePagingIterator.hasNext();
+    }
+
+    @Override
+    public Row next() {
+        return delegatePagingIterator.next();
+    }
+
+    @Override
+    public void remove() {
+        delegatePagingIterator.remove();
+    }
+
+    @Override
+    public void forEachRemaining(Consumer<? super Row> action) {
```

Comment:
The default implementation calls into `hasNext()` / `next()` and those calls will be forwarded to the delegate